libraries/architecture/src/main/kotlin/io/element/android/libraries/architecture/Async.kt

```diff
@@ -18,51 +18,148 @@ package io.element.android.libraries.architecture
 
 import androidx.compose.runtime.MutableState
 import androidx.compose.runtime.Stable
+import kotlin.contracts.ExperimentalContracts
+import kotlin.contracts.InvocationKind
+import kotlin.contracts.contract
 
+/**
+ * Sealed type that allows to model an asynchronous operation.
+ */
 @Stable
 sealed interface Async<out T> {
+
+    /**
+     * Represents a failed operation.
+     *
+     * @param exception the exception that caused the operation to fail.
+     * @param prevData the data returned by a previous successful run of the operation if any.
+     */
+    data class Failure<out T>(
+        val exception: Throwable,
+        val prevData: T? = null,
+    ) : Async<T>
+
+    /**
+     * Represents an operation that is currently ongoing.
+     *
+     * @param prevData the data returned by a previous successful run of the operation if any.
+     */
+    data class Loading<out T>(
+        val prevData: T? = null,
+    ) : Async<T>
+
+    /**
+     * Represents a successful operation.
+     *
+     * @param data the data returned by the operation.
+     */
+    data class Success<out T>(
+        val data: T,
+    ) : Async<T>
+
+    /**
+     * Represents an uninitialized operation (i.e. yet to be run).
+     */
     object Uninitialized : Async<Nothing>
-    data class Loading<out T>(val prevState: T? = null) : Async<T>
-    data class Failure<out T>(val error: Throwable, val prevState: T? = null) : Async<T>
-    data class Success<out T>(val state: T) : Async<T>
-
-    fun dataOrNull(): T? {
-        return when (this) {
-            is Failure -> prevState
-            is Loading -> prevState
-            is Success -> state
-            Uninitialized -> null
-        }
+
+    /**
+     * Returns the data returned by the operation, or null otherwise.
+     *
+     * Please note this method may return stale data if the operation is not [Success].
+     */
+    fun dataOrNull(): T? = when (this) {
+        is Failure -> prevData
+        is Loading -> prevData
+        is Success -> data
+        Uninitialized -> null
     }
-}
 
-suspend inline fun <T> (suspend () -> T).execute(
-    state: MutableState<Async<T>>,
-    errorMapping: ((Throwable) -> Throwable) = { it },
-) {
-    try {
-        state.value = Async.Loading()
-        val result = this()
-        state.value = Async.Success(result)
-    } catch (error: Throwable) {
-        state.value = Async.Failure(errorMapping.invoke(error))
+    /**
+     * Returns the exception that caused the operation to fail, or null otherwise.
+     */
+    fun exceptionOrNull(): Throwable? = when (this) {
+        is Failure -> exception
+        else -> null
     }
+
+    fun isFailure(): Boolean = this is Failure<T>
+
+    fun isLoading(): Boolean = this is Loading<T>
+
+    fun isSuccess(): Boolean = this is Success<T>
+
+    fun isUninitialized(): Boolean = this == Uninitialized
```

Comment:
This could be useful too: https://github.com/airbnb/mavericks/blob/main/mvrx-common/src/main/java/com/airbnb/mvrx/Async.kt#L86C21-L86C21

Reply:
Yeah indeed. Let's leave it out for this first iteration. It's a non breaking change that can be added at any time.